File: server/services/backupService.js
const fs = require('fs').promises;
const path = require('path');
const archiver = require('archiver');
const mongoose = require('mongoose');
const Backup = require('../models/Backup');
const Product = require('../models/Product');
const PackingList = require('../models/PackingList');

class BackupService {
  constructor() {
    this.backupDir = path.join(__dirname, '../backups');
  }

  async init() {
    try {
      await fs.mkdir(this.backupDir, { recursive: true });
    } catch (error) {
      console.error('创建备份目录失败:', error);
      throw error;
    }
  }

  async createBackup(type, userId) {
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const filename = `backup-${type}-${timestamp}.zip`;
    const filepath = path.join(this.backupDir, filename);

    const backup = await Backup.create({
      filename,
      size: 0,
      type,
      path: filepath,
      createdBy: userId
    });

    try {
      const output = fs.createWriteStream(filepath);
      const archive = archiver('zip', {
        zlib: { level: 9 }
      });

      archive.pipe(output);

      // 根据类型备份不同的数据
      switch (type) {
        case 'products':
          await this.backupProducts(archive);
          break;
        case 'packingLists':
          await this.backupPackingLists(archive);
          break;
        case 'full':
          await this.backupProducts(archive);
          await this.backupPackingLists(archive);
          break;
      }

      await archive.finalize();

      const stats = await fs.stat(filepath);
      backup.size = stats.size;
      backup.status = 'completed';
      backup.completedAt = new Date();
      await backup.save();

      return backup;
    } catch (error) {
      backup.status = 'failed';
      backup.error = error.message;
      await backup.save();
      throw error;
    }
  }

  async backupProducts(archive) {
    const products = await Product.find({});
    archive.append(JSON.stringify(products, null, 2), { name: 'products.json' });
  }

  async backupPackingLists(archive) {
    const packingLists = await PackingList.find({});
    archive.append(JSON.stringify(packingLists, null, 2), { name: 'packingLists.json' });
  }

  async restoreBackup(backupId) {
    const backup = await Backup.findById(backupId);
    if (!backup) {
      throw new Error('备份不存在');
    }

    const fileContent = await fs.readFile(backup.path);
    const data = JSON.parse(fileContent);

    const session = await mongoose.startSession();
    try {
      await session.withTransaction(async () => {
        switch (backup.type) {
          case 'products':
            await Product.deleteMany({}, { session });
            await Product.insertMany(data.products, { session });
            break;
          case 'packingLists':
            await PackingList.deleteMany({}, { session });
            await PackingList.insertMany(data.packingLists, { session });
            break;
          case 'full':
            await Product.deleteMany({}, { session });
            await PackingList.deleteMany({}, { session });
            await Product.insertMany(data.products, { session });
            await PackingList.insertMany(data.packingLists, { session });
            break;
        }
      });
    } finally {
      await session.endSession();
    }
  }

  async listBackups(query = {}) {
    return Backup.find(query)
      .sort({ createdAt: -1 })
      .populate('createdBy', 'username');
  }

  async deleteBackup(backupId) {
    const backup = await Backup.findById(backupId);
    if (!backup) {
      throw new Error('备份不存在');
    }

    try {
      await fs.unlink(backup.path);
    } catch (error) {
      console.error('删除备份文件失败:', error);
    }

    await backup.deleteOne();
  }
}

module.exports = new BackupService(); 

File: backend/app/services/backup_service.py
import os
import json
import shutil
from datetime import datetime
from typing import List, Optional
from sqlalchemy.orm import Session
import aiofiles
from fastapi import UploadFile

from ..models.backup import Backup, BackupType, BackupStatus
from ..models.product import Product
from ..models.packing_list import PackingList
from ..config import settings

class BackupService:
    def __init__(self):
        self.backup_dir = os.path.join(os.getcwd(), "backups")
        os.makedirs(self.backup_dir, exist_ok=True)

    async def create_backup(
        self,
        db: Session,
        backup_type: BackupType,
        user_id: int,
        description: Optional[str] = None
    ) -> Backup:
        """创建备份"""
        timestamp = datetime.utcnow().strftime("%Y%m%d_%H%M%S")
        filename = f"backup_{backup_type}_{timestamp}.json"
        filepath = os.path.join(self.backup_dir, filename)

        # 创建备份记录
        backup = Backup(
            filename=filename,
            size=0,
            type=backup_type,
            path=filepath,
            description=description,
            created_by=user_id
        )
        db.add(backup)
        db.commit()

        try:
            # 根据类型备份不同的数据
            data = {}
            if backup_type == BackupType.PRODUCTS or backup_type == BackupType.FULL:
                products = db.query(Product).all()
                data["products"] = [product.__dict__ for product in products]

            if backup_type == BackupType.PACKING_LISTS or backup_type == BackupType.FULL:
                packing_lists = db.query(PackingList).all()
                data["packing_lists"] = [pl.__dict__ for pl in packing_lists]

            # 写入备份文件
            async with aiofiles.open(filepath, "w", encoding="utf-8") as f:
                await f.write(json.dumps(data, ensure_ascii=False, indent=2))

            # 更新备份状态
            file_size = os.path.getsize(filepath)
            backup.size = file_size
            backup.status = BackupStatus.COMPLETED
            backup.completed_at = datetime.utcnow()
            db.commit()

            return backup

        except Exception as e:
            backup.status = BackupStatus.FAILED
            backup.error = str(e)
            db.commit()
            raise

    async def restore_backup(
        self,
        db: Session,
        backup_id: int
    ) -> None:
        """恢复备份"""
        backup = db.query(Backup).filter(Backup.id == backup_id).first()
        if not backup:
            raise ValueError("备份不存在")

        if backup.status != BackupStatus.COMPLETED:
            raise ValueError("备份未完成，无法恢复")

        try:
            # 读取备份文件
            async with aiofiles.open(backup.path, "r", encoding="utf-8") as f:
                content = await f.read()
                data = json.loads(content)

            # 根据备份类型恢复数据
            if backup.type in [BackupType.PRODUCTS, BackupType.FULL]:
                # 清空现有数据
                db.query(Product).delete()
                # 恢复产品数据
                for product_data in data.get("products", []):
                    product = Product(**product_data)
                    db.add(product)

            if backup.type in [BackupType.PACKING_LISTS, BackupType.FULL]:
                # 清空现有数据
                db.query(PackingList).delete()
                # 恢复装箱单数据
                for pl_data in data.get("packing_lists", []):
                    packing_list = PackingList(**pl_data)
                    db.add(packing_list)

            db.commit()

        except Exception as e:
            db.rollback()
            raise ValueError(f"恢复备份失败: {str(e)}")

    def list_backups(
        self,
        db: Session,
        skip: int = 0,
        limit: int = 10,
        backup_type: Optional[BackupType] = None
    ) -> List[Backup]:
        """获取备份列表"""
        query = db.query(Backup)
        if backup_type:
            query = query.filter(Backup.type == backup_type)
        return query.order_by(Backup.created_at.desc()).offset(skip).limit(limit).all()

    async def delete_backup(
        self,
        db: Session,
        backup_id: int
    ) -> None:
        """删除备份"""
        backup = db.query(Backup).filter(Backup.id == backup_id).first()
        if not backup:
            raise ValueError("备份不存在")

        try:
            # 删除备份文件
            if os.path.exists(backup.path):
                os.remove(backup.path)

            # 删除数据库记录
            db.delete(backup)
            db.commit()

        except Exception as e:
            db.rollback()
            raise ValueError(f"删除备份失败: {str(e)}")

backup_service = BackupService() 